# InstaCart Project 

## Ex4.9 Part 1-  Import, wrangle and integrate Customer data with main dataframe

### Contents
- 1. Import libraries and new customer data
- 2. Exploring customer dataframe
- 3.  Looking for missing values
- 4. Rename columns
- 5. Change data type of user_id
- 6. Correcting columns containing mixed data types
- 7. Checking for duplicates 
- 8. Merging customer df with ords_prods_merge df
	-  Checking for common column and verifying datatype
	- Import ords_prods_merge df
	- Check common key column is same datatype as that in customer df
	- Test merge and show indicator key
- 10. Export  new merged df as ‘orders_products_customers_merged.pkl’

# 01 Import libraries and new customer data

In [1]:
# Importing libraries
import pandas as pd 
import numpy as np 
import os 

In [2]:
# Create file path
path = r'//Users/amypalomino/Documents/(03.22) InstaCart Basket Analysis'

In [4]:
# Import customer dataframe 
customer = pd.read_csv(
        os.path.join(path,'02 Data', 'Original Data', 'customers.csv'), 
        index_col = False)

# 02  Exploring customer dataframe

In [5]:
customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
customer.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

In [7]:
customer.shape

(206209, 10)

In [8]:
customer.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [34]:
customer.info

<bound method DataFrame.info of         user_id First Name   Surname  Gender           State  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020

In [35]:
customer.dtypes

user_id           int64
First Name       object
Surname          object
Gender           object
State            object
Age               int64
date_joined      object
n_dependants      int64
Family_status    object
income            int64
dtype: object

In [9]:
customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

# 03 Looking for missing values 

In [15]:
customer['Gender'].value_counts(dropna = False)

Male      104067
Female    102142
Name: Gender, dtype: int64

In [17]:
customer[['First Name', 'Surnam']].value_counts(dropna = False)

First Name  Surnam    
Andrea      Hamilton      252
Ralph       Randall       248
Shawn       Lamb          243
Brenda      Pennington    243
Linda       Barnett       242
                         ... 
Gloria      Bauer         166
Marilyn     Pearson       164
Irene       Payne         163
Pamela      Jordan        162
Betty       Bean            1
Length: 1001, dtype: int64

- This frequency count of first and last name has possibly given me the number of transactions made by each customer? 
- Or is it the number of times those names appear in the dataframe?

### Will a frequency count of the user_id show 206,209 unique records?

In [19]:
customer['user_id'].value_counts(dropna = False)

26711     1
67322     1
173044    1
61044     1
98344     1
         ..
146847    1
154991    1
172193    1
184326    1
80148     1
Name: user_id, Length: 206209, dtype: int64

In [18]:
customer['STATE'].value_counts(dropna = False)

Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana       

- This result has suprised me - why are there equal numbers of records for every single state in the dataframe? 
- Does this indicate that there an even number of customers in every state!?

In [21]:
customer['Age'].value_counts(dropna = False)

19    3329
55    3317
51    3317
56    3306
32    3305
      ... 
65    3145
25    3127
66    3114
50    3102
36    3101
Name: Age, Length: 64, dtype: int64

In [23]:
customer.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [26]:
customer_nan = customer[customer['First Name'].isnull() == True]
customer_nan

user_id First Name    Surnam  Gender         STATE  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

- I've established that the only missing value is in the Furst Name column. 
- I'm going to leave this columns as it is as the remaining 95% of the df contains first names


In [31]:
customer['fam_status'].value_counts(dropna = False)

married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: fam_status, dtype: int64

# 04 Rename columns

In [27]:
# To correct the typo in the Surname column
customer.rename(columns = {'Surnam' : 'Surname'}, inplace = True)

In [29]:
# Change STATE to lowercase
customer.rename(columns = {'STATE' : 'State'}, inplace = True)

In [36]:
# Rename n_dependants to No_of_dependants
customer.rename(columns = {'n_dependants' : 'No_of_dependants'}, inplace = True)

In [32]:
# Rename fam_status to Family_status
customer.rename(columns = {'fam_status' : 'Family_status'}, inplace = True)

In [37]:
customer.head()

user_id First Name   Surname  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   No_of_dependants Family_status  income  
0                 3       married  165665  
1                 0        single   59285  
2                 2       married   99568  
3                 0        single   42049  
4                 1       married   40374

# 05 Change data type of user_id

In [39]:
# To improve efficiency, change uder_id to str data type rather than int64
customer['user_id'] = customer['user_id'].astype('str')

In [42]:
customer.dtypes

user_id             object
First Name          object
Surname             object
Gender              object
State               object
Age                  int64
date_joined         object
No_of_dependants     int64
Family_status       object
income               int64
dtype: object

# 06 Correcting columns containing mixed data types


In [43]:
for col in customer.columns.tolist():
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer[weird]) > 0:
    print (col)

First Name


- First Name column appears to contain mixed data types. I already know it also contains missing values. I'll change data trype to string

In [44]:
customer['First Name'] = customer['First Name'].astype('str')

In [45]:
# Check to see if column only contains one data type
for col in customer.columns.tolist():
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer[weird]) > 0:
    print (col)

- Nothing was printed which means this is now resolved. 

# 07 Checking for duplicates 


In [46]:
customer_dups = customer[customer.duplicated()]

In [47]:
customer_dups

Empty DataFrame
Columns: [user_id, First Name, Surname, Gender, State, Age, date_joined, No_of_dependants, Family_status, income]
Index: []

- No duplicates were returned here

# 08 Merging customer df with ords_prods_merge df

### Checking for common column and verifying datatype

In [48]:
customer.head()

user_id First Name   Surname  Gender       State  Age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   No_of_dependants Family_status  income  
0                 3       married  165665  
1                 0        single   59285  
2                 2       married   99568  
3                 0        single   42049  
4                 1       married   40374

In [51]:
customer.dtypes['user_id']

dtype('O')

### Import ords_prods_merge df

In [53]:
# Import orders_products df as 'ords_prods'
ords_prods = pd.read_pickle(
        os.path.join(path,'02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [54]:
ords_prods.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   ... prices    price_range_loc   busiest day  busiest_period_of_day  \
0  ...    9.0  Mid range product  Regular Busy         Average orders   
1  ...   12.5  Mid range product  Regular Busy         Average orders   
2  ...    4.4  Low range product  Regular Busy         Average orders   
3  ...    4.7  Low range product  Regular Busy         Average orders   
4  ...    1.0  Low range product  Regular Busy         Average orders   

  max_order  loyalty_flag average_price  High/Low_Spender  \
0        10  New customer      6.367797       Low Spender   
1        10  New customer      6.367797       Low Spender   
2        10  New customer      6.367797       Low Spender   
3        10  New customer      6.367797       Low Spender   
4        10  New customer      6.367797       Low Spender   

  median_order_frequency   Order_frequency_flag  
0                   20.5  Non-frequent customer  
1                   20.5  Non-frequent customer  
2                   20.5  Non-frequent customer  
3                   20.5  Non-frequent customer  
4                   20.5  Non-frequent customer  

[5 rows x 23 columns]

### Check common key column is same datatype as that in customer df

In [56]:
ords_prods.dtypes['user_id']

dtype('int64')

- Unfortuantely not. I'll change the user_id datatype to 'str' in the ords_prods df to match the customer df user_id datatype 

In [57]:
ords_prods['user_id'] = ords_prods['user_id'].astype('str')

In [58]:
ords_prods.dtypes['user_id']

dtype('O')

- The common key that I'll use to merge the 2 dfs is now the same datatype

In [59]:
ords_prods.shape

(32435059, 23)

### Test merge and show indicator key

In [61]:
# Test a left join merge

pd.merge(ords_prods, customer, on = ['user_id'], how = 'left')

order_id user_id  order_number  order_day_of_week  \
0          2539329       1             1                  2   
1          2539329       1             1                  2   
2          2539329       1             1                  2   
3          2539329       1             1                  2   
4          2539329       1             1                  2   
...            ...     ...           ...                ...   
32435054   2977660  206209            13                  1   
32435055   2977660  206209            13                  1   
32435056   2977660  206209            13                  1   
32435057   2977660  206209            13                  1   
32435058   2977660  206209            13                  1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32435054                 12                     7.0       14197   
32435055                 12                     7.0       38730   
32435056                 12                     7.0       31477   
32435057                 12                     7.0        6567   
32435058                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  ...   Order_frequency_flag  \
0                         1          0   both  ...  Non-frequent customer   
1                         2          0   both  ...  Non-frequent customer   
2                         3          0   both  ...  Non-frequent customer   
3                         4          0   both  ...  Non-frequent customer   
4                         5          0   both  ...  Non-frequent customer   
...                     ...        ...    ...  ...                    ...   
32435054                  5          1   both  ...  Non-frequent customer   
32435055                  6          0   both  ...  Non-frequent customer   
32435056                  7          0   both  ...  Non-frequent customer   
32435057                  8          0   both  ...  Non-frequent customer   
32435058                  9          0   both  ...  Non-frequent customer   

          First Name  Surname  Gender    State Age date_joined  \
0              Linda   Nguyen  Female  Alabama  31   2/17/2019   
1              Linda   Nguyen  Female  Alabama  31   2/17/2019   
2              Linda   Nguyen  Female  Alabama  31   2/17/2019   
3              Linda   Nguyen  Female  Alabama  31   2/17/2019   
4              Linda   Nguyen  Female  Alabama  31   2/17/2019   
...              ...      ...     ...      ...  ..         ...   
32435054       Diana  Donovan  Female     Iowa  74   9/14/2019   
32435055       Diana  Donovan  Female     Iowa  74   9/14/2019   
32435056       Diana  Donovan  Female     Iowa  74   9/14/2019   
32435057       Diana  Donovan  Female     Iowa  74   9/14/2019   
32435058       Diana  Donovan  Female     Iowa  74   9/14/2019   

          No_of_dependants Family_status  income  
0                        3       married   40423  
1                        3       married   40423  
2                        3       married   40423  
3                        3       married   40423  
4                        3       married   40423  
...                    ...           ...     ...  
32435054                 3       married  137969  
32435055                 3       married  137969  
32435056                 3       married  137969  
32435057                 3       married  137969  
32435058                 3       married  137969  

[32435059 rows x 32 columns]

In [62]:
# Happy with test merge so will create new df called 'ords_prods_custs', merging dfs on left join
ords_prods_custs = ords_prods.merge(customer, on = ['user_id'], how = 'left')

In [63]:
ords_prods_custs.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2539329       1             1                  2                  8   
2   2539329       1             1                  2                  8   
3   2539329       1             1                  2                  8   
4   2539329       1             1                  2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   ...   Order_frequency_flag  First Name  Surname  Gender    State Age  \
0  ...  Non-frequent customer       Linda   Nguyen  Female  Alabama  31   
1  ...  Non-frequent customer       Linda   Nguyen  Female  Alabama  31   
2  ...  Non-frequent customer       Linda   Nguyen  Female  Alabama  31   
3  ...  Non-frequent customer       Linda   Nguyen  Female  Alabama  31   
4  ...  Non-frequent customer       Linda   Nguyen  Female  Alabama  31   

  date_joined  No_of_dependants Family_status  income  
0   2/17/2019                 3       married   40423  
1   2/17/2019                 3       married   40423  
2   2/17/2019                 3       married   40423  
3   2/17/2019                 3       married   40423  
4   2/17/2019                 3       married   40423  

[5 rows x 32 columns]

In [64]:
ords_prods_custs.shape

(32435059, 32)

In [65]:
ords_prods_custs.value_counts('_merge')

_merge
both          32435059
left_only            0
right_only           0
dtype: int64

# 10 Export new merged dataframe as pkl

In [66]:
ords_prods_custs.to_pickle(os.path.join(
    path, '02 Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))